examples/mqtt_demo/micropython/micropython_agent.py

```diff
@@ -0,0 +1,247 @@
+                    "args": {
+                        "original_message": message,
+                        "error": f"{e}",
+                    },
+                }
+            )
+
+    def __commit(self, message: dict):
+        """
+        Invokes action if permitted otherwise raises PermissionError
+        """
+        # Check if the action method exists
+        action_method = None
+        try:
+            action_method = getattr(
+                self, f"{self.ACTION_METHOD_PREFIX}{message['action']}"
+            )
+        except AttributeError as e:
+            # the action was not found
+            if message["to"] == self.id():
+                # if it was point to point, raise an error
+                raise AttributeError(
+                    f"\"{message['action']}\" action not found on \"{self.id()}\""
+                )
+            else:
+                # broadcasts will not raise an error
+                return
+
+        self._before_action(message)
+
+        return_value = None
+        error = None
+        try:
+            # Check if the action is permitted
+            if self.__permitted(message):
+                # Invoke the action method
+                # (set _current_message so that it can be used by the action)
+                self._current_message = message
+                return_value = action_method(**message["args"])
+                self._current_message = None
+
+                # An immediate data structure response (return value) if any, from an
+                # action method is sent back to the sender as a "return" action. This is
+                # useful for actions that simply need to return a value to the sender.
+                if return_value is not None:
+                    self._send(
+                        {
+                            "to": message["from"],
+                            "thoughts": "A value was returned for your action",
+                            "action": "return",
+                            "args": {
+                                "original_message": message,
+                                "return_value": return_value,
+                            },
+                        }
+                    )
+            else:
+                raise PermissionError(
+                    f"\"{self.id()}.{message['action']}\" not permitted"
+                )
+        except Exception as e:
+            error = e  # save the error for _after_action
+            raise Exception(e)
+        finally:
+            self._after_action(message, return_value, error)
+
+    def __permitted(self, message: dict) -> bool:
+        """
+        Checks whether the action represented by the message is allowed
+        """
+        # policy = getattr(self, f"{self.ACTION_METHOD_PREFIX}{message['action']}").access_policy
+        policy = _function_access_policies[f"{self.ACTION_METHOD_PREFIX}{message['action']}"][
+            "access_policy"
+        ]
```

Comment:
It's always a good idea to try not to duplicate code.

I am currently encountering a difficulty. There are differences in the design philosophy of CPython and MicroPython, serving different goals. The convenience of CPython is great, there are many batteries available out of the box. 

Since MicroPython runs in a hardware environment with limited resources, one of its design goals is to keep it small and concise.

If we want to make the base class compatible with CPython and MicroPython, we may need a more basic base class, 
Like [Space base class](https://github.com/operand/agency/blob/main/agency/space.py), more focus on interface contracts. Then let both CPython and MicroPython inherit from it. 

Do you agree with this refactoring?

Reply:
yes I agree. Maybe we can name both base classes `AbstractSpace` and `AbstractAgent` to make it clearer. And we can reserve the `Agent` class name for the CPython implementation since it'll be used most often.

If this refactoring is too painful right now it's fine to merge without it and we can do the refactoring as a separate task.

Reply:
The following is the part of the Agent base class that is not compatible with MicroPython:

- import inspect: There is an inspect library in micropython, but its function is extremely weak 
- import threading:  A low-level threading library exists in micropython: [_thread](https://docs.micropython.org/en/latest/library/_thread.html), It is missing Event.
- from colorama import Fore, Style 
- from agency.schema import MessageSchema: It seems that we can use assert for a simple check: `assert message["from"] in message`

Reply:
> Maybe we can name both base classes AbstractSpace and AbstractAgent to make it clearer. And we can reserve the Agent class name for the CPython implementation since it'll be used most often.

we happen to think the same!

> If this refactoring is too painful right now it's fine to merge without it and we can do the refactoring separately.

I just tried  and it's easier than I thought. I'm would like to try it, I'll make a commit later, feel free to suggest improvements!

Reply:
Some thoughts on what we can do:
- `inspect` - Maybe for the MicroPython implementation we explicitly require that the `_help` method be implemented? It could be an abstract method in the base class and we keep the current implementation in the CPython `Agent` class.
- `threading` - You know, it's actually a bit dirty that those events are declared in the `Agent` class. We could maintain those events entirely within the `Space` class. This keeps all the threading code in just one place. I probably should've done it that way.
- `colorama` - We should probably remove this dependency. We can make the text yellow by just using the [ansi color codes](https://talyian.github.io/ansicolors/) directly.
- `MessageSchema` - This one I'm not sure of. If we're going to create an abstract base class, then I think we can use that approach in the MicroPython implementation. I do like having the validation done with pydantic but we can live without it if we have to.